kafka-connect-fitbit-source/src/main/java/org/radarbase/connect/rest/fitbit/route/FitbitPollingRoute.java

```diff
@@ -0,0 +1,329 @@
+    offsets.put(userKey, offset);
+  }
+
+  @Override
+  public void requestEmpty(RestRequest request) {
+    FitbitRestRequest fitbitRequest = (FitbitRestRequest) request;
+    Instant endOffset = fitbitRequest.getDateRange().end().toInstant();
+    if (DAYS.between(endOffset, lastPoll) >= HISTORICAL_TIME_DAYS) {
+      String key = fitbitRequest.getUser().getId();
+      offsets.put(key, endOffset);
+    }
+  }
+
+  @Override
+  public void requestFailed(RestRequest request, Response response) {
+    if (response != null && response.code() == 429) {
+      FitbitUser user = ((FitbitRestRequest)request).getUser();
+      logger.info("Too many requests for user {}. Backing off for {}",
+          user, TOO_MANY_REQUESTS_COOLDOWN);
+      lastPollPerUser.compute(user.getId(), (u, p) -> p == null
+          ? lastPoll.plus(TOO_MANY_REQUESTS_COOLDOWN)
+          : PollingRequestRoute.max(
+              p.plus(getPollIntervalPerUser()), lastPoll.plus(TOO_MANY_REQUESTS_COOLDOWN)));
+    } else {
+      logger.warn("Failed to make request {}", request);
+    }
+  }
+
+  /**
+   * Actually construct requests, based on the current offset
+   * @param user Fitbit user
+   * @return request to make
+   */
+  protected abstract Stream<FitbitRestRequest> createRequests(FitbitUser user);
+
+  @Override
+  public Stream<FitbitRestRequest> requests() {
+    lastPoll = Instant.now();
+    try {
+      return userRepository.stream()
+          .map(u -> new AbstractMap.SimpleImmutableEntry<>(u, nextPoll(u)))
+          .filter(u -> lastPoll.isAfter(u.getValue()))
+          .sorted(Comparator.comparing(Map.Entry::getValue))
+          .limit(maxUsersPerPoll)
+          .flatMap(u -> {
+            lastPollPerUser.put(u.getKey().getId(), lastPoll);
+            return this.createRequests(u.getKey());
+          })
+          .filter(Objects::nonNull);
+    } catch (IOException e) {
+      logger.warn("Cannot read users");
+      return Stream.empty();
+    }
+  }
+
+  private Map<String, Object> getPartition(FitbitUser user) {
+    return partitions.computeIfAbsent(user.getId(),
+        k -> generator.getPartition(routeName, user));
+  }
+
+  /**
+   * Create a FitbitRestRequest for given arguments.
+   * @param user Fitbit user
+   * @param dateRange dates that may be queried in the request
+   * @param urlFormatArgs format arguments to {@link #getUrlFormat(String)}.
+   * @return request or {@code null} if the authorization cannot be arranged.
+   */
+  protected FitbitRestRequest newRequest(FitbitUser user, DateRange dateRange,
+      Object... urlFormatArgs) {
+    Request.Builder builder = new Request.Builder()
+        .url(String.format(getUrlFormat(baseUrl), urlFormatArgs));
+    try {
+      Request request = builder
+          .header("Authorization", "Bearer " + userRepository.getAccessToken(user))
+          .build();
+      return new FitbitRestRequest(this, request, user, getPartition(user),
```

Comment:
I am wondering whether this is redundant with what `TokenAuthenticator` would do? In other words when either of these are used?

Reply:
TokenAuthenticator is only invoked if authentication fails (i.e. no credentials or expired credentials were given). This adds authentication pro-actively to avoid the first round-trip.